include/zephyr/drivers/sensor_types.h

```diff
@@ -0,0 +1,264 @@
+
+/**
+ * Ambient light sensor returning data in lux.
+ *
+ * This is an on-change sensor.
+ */
+#define SENSOR_TYPE_LIGHT UINT8_C(12)
+
+/**
+ * Proximity detection.
+ *
+ * This is an on-change sensor.
+ */
+#define SENSOR_TYPE_PROXIMITY UINT8_C(13)
+
+/**
+ * Step detection.
+ */
+#define SENSOR_TYPE_STEP_DETECT UINT8_C(23)
+
+/**
+ * Step counter.
+ *
+ * This is an on-change sensor.
+ */
+#define SENSOR_TYPE_STEP_COUNTER UINT8_C(24)
+
+/**
+ * Accelerometer temperature sensor returning the accelerometer's temperature in Kelvin.
+ */
+#define SENSOR_TYPE_ACCELEROMETER_TEMPERATURE UINT8_C(56)
+
+/**
+ * Gyroscope temperature sensor returning the gyroscope's temperature in Kelvin.
+ */
+#define SENSOR_TYPE_GYROSCOPE_TEMPERATURE UINT8_C(57)
+
+/**
+ * Magnetometer temperature sensor returning the magnetometer's temperature in Kelvin.
+ */
+#define SENSOR_TYPE_GEOMAGNETIC_FIELD_TEMPERATURE UINT8_C(58)
+
+BUILD_ASSERT(SENSOR_TYPE_GEOMAGNETIC_FIELD_TEMPERATURE < SENSOR_TYPE_VENDOR_START,
+	     "Too many sensor types");
+
+struct sensor_data_header {
+	/** The number of allocated `readings` elements. */
+	const uint16_t reading_size;
+
+	/** The base timestamp, in nanoseconds. */
+	uint64_t base_timestamp;
+
+	/** The number of valid elements in the 'readings' array. */
+	uint16_t reading_count;
+};
+
+/**
+ * @brief A set of 3D data points from the sensor.
+ *
+ * This is used by:
+ * - SENSOR_TYPE_ACCELEROMETER
+ * - SENSOR_TYPE_GYROSCOPE
+ * - SENSOR_TYPE_MAGNETOMETER
+ * - SENSOR_TYPE_GEOMAGNETIC_FIELD
+ */
+struct sensor_three_axis_data {
+	/** Header data */
+	struct sensor_data_header header;
+
+	/** The data from the sensor */
+	struct sensor_three_axis_sample_data {
+		union {
+			uint32_t raw_values[3];
+			struct {
+				uint32_t raw_x;
+				uint32_t raw_y;
+				uint32_t raw_z;
+			};
+			fp_t values[3];
+			struct {
+				fp_t x;
+				fp_t y;
+				fp_t z;
+			};
+		};
+	} readings[1];
+};
+
+/**
+ * @brief A raw single byte data points from the sensor.
+ *
+ * This is used by:
+ * - SENSOR_TYPE_PROXIMITY
+ */
+struct sensor_byte_data {
+	/** Header data */
+	struct sensor_data_header header;
+
+	/** The data from the sensor */
+	struct sensor_byte_sample_data {
+		union {
+			uint8_t value;
+			struct {
+				uint8_t is_near : 1;
+				uint8_t padding : 7;
+			};
+		};
+	} readings[1];
+};
+
+/**
+ * This is used by:
+ * - SENSOR_TYPE_LIGHT
+ * - SENSOR_TYPE_PRESSURE
+ * - SENSOR_TYPE_ACCELEROMETER_TEMPERATURE
+ * - SENSOR_TYPE_GYROSCOPE_TEMPERATURE
+ * - SENSOR_TYPE_MAGNETIC_FIELD_TEMPERATURE
+ */
+struct sensor_float_data {
+	struct sensor_data_header header;
+	struct sensor_float_sample_data {
+		union {
+			uint32_t raw_value;
+			fp_t value;
+		};
+	} readings[1];
+};
+
+/**
+ * This is used by:
+ * - SENSOR_TYPE_STEP_COUNTER
+ */
+struct sensor_uint64_data {
+	struct sensor_data_header header;
+	struct sensor_uint64_sample_data {
+		uint64_t value;
+	} readings[1];
+};
+
+struct sensor_raw_data {
```

Comment:
I haven't finished reviewing the BMI160 implementation of this API, but is the intention that the user can/will cast pointers to these sensor data objects from one type to another? That would be a good thing to call out in this header.

More broadly, while the API documents the individual functions and types, I'm finding it hard to get a sense of how they are supposed to be used without referring to the implementation. Ideally, these headers would be more approachable.